README.md
# Handyplex
Handyplex allows you to control Plex via the XBox Kinect hardware. Using
simple hand gestures you can navigate through your library and initiate
playback commands such as Play/Pause, Volume Up/Down and Fast-Forward/Rewind.

See Handyplex in action: http://www.youtube.com/watch?v=wFM7pVMLIY4

Handyplex is currently a proof of concept and thus far has been tested
successfully on Mac OS X 10.6.8 (Snow Leopard) running Plex 0.9.3.4, and
Mac OS X 10.7 (Lion) running Plex 0.9.5.2 (Laika). Since all used libraries are
cross-platform, it should run on other platforms too.

Due to the experimental state of Handyplex one needs to be adept with the command
line to build and install dependencies, but hopefully the following instructions
will help you along.

## Requirements
To use Handyplex you'll need the following:

* Mac OS X Snow Leopard (other platforms might work, depending on
  OpenNI/Nite support)
* Python 2.7
* Plex (tested with 0.9.3.4 and 0.9.5.2)
* OSCeleton and it's dependencies

Hardware:

* Microsoft Xbox Kinect Camera

## Installing
The following is an abridged version of the installation instructions,
see the supplied INSTALL file for step-by-step instructions.

Make sure Python 2.7 (or later) is installed correctly. I recommend
[virtualenv](http://pypi.python.org/pypi/virtualenv) and [pip](http://www.pip-installer.org) to set up your Python environments.


1. Install OSCeleton and follow the instructions to install the OpenNI/NITE
   dependencies: [OSCeleton](https://github.com/Sensebloom/OSCeleton)
1. Clone Handyplex
1. Install Handyplex's Python dependencies. The easiest way to do this
   is to use pip with the supplied requirements.txt:

        $ pip install -r requirements.txt

   See [pip](http://www.pip-installer.org/en/latest/index.html) for more information.

(Optional) Install [wxPython](http://www.wxpython.org/) to play sounds when users are detected/lost.

## Configuring
Handyplex needs to know the IP where Plex Media Server is running, as well as
the name of the computer on which the Plex client is running.
Edit the following lines of settings.py accordingly:

    PLEX_SERVER_IP = '10.0.1.1'
    PLEX_CLIENT_NAME = 'mini.local'

For Handyplex to start OSCeleton you'll need to specify the directory where
the executable resides:

    OSCELETON_PATH = '/Users/mini/Development/OSCeleton'


## Using Handyplex
Make sure Plex is running, and the Kinect is connected via USB and is
receiving power.

1. Run Handyplex:

        $ python handyplex.py

1. Initiate gesture recognition with a 'wave' gesture. You should hear a
   sound letting you know your hand is recognized (if you have wxPython 
   installed), and should be good to go.


### Gesture Recognition
To get a feel for how the gesture recognition works see the demo video
on Youtube http://www.youtube.com/watch?v=wFM7pVMLIY4

#### Wave Gesture
To initiate gesture recognition start with a 'wave' gesture. Move your
hand on a horizontal axis 3 or 4 times, as if you are waving to
somebody.

#### Hold Gesture
Hold gestures are necessary to determine the start point for further
gesture recognition. Simply hold your hand in one place for a short
amount of time.

#### Swipe Gestures
Swipe gestures are best recognized if performed in a short motion, either
left, right, up, down, forwards or backwards. To implement a flick,
perform a swipe gesture and quickly move your hand back to the starting
position.

#### Repeating Swipe Gestures
If you 'hold' a gesture after it has been detected it
Handyplex will recognize 'repeat' actions, repeatedly performing the
last detected action. Repeat actions can be accelerated/decelerated by
moving your hand further away or closer to the gesture recognition
point. Only movements in the same direction as the original gesture are
considered to be repeat movements, moving your hand too far off the
original axis cancels the repeat action.

#### Fine-tuning
If you feel gesture recognition isn't working as smoothly as in the
demo video you might want to play around with the different thresholds
in settings.py.


## Running Unit Tests
Either run the tests individually, or use unittest test discover
functionality (run in tests directory):

    $ python -m unittest discover


## Known Issues
* Since there currently is no 'stop gesture recognition' command, you need to
  hide your hand (behind your back for example) in order to stop gesture
  recognition.
* Sometimes a gesture is not recognized with a small movements, move your
  hand in a exaggerated motion to one direction to reset.

## Future Plans
* Implement a gesture for ending further gesture recognition until
  user re-initializes session with 'wave'.
* Have PlexController send key commands instead of using the HTTP API
  for more flexibility.
* Use NITE's gesture recognition instead of the current Python
  implementation.
* Support for multi-hand gestures.

## Feedback
All feedback is very welcome.
Please use the github issue tracker to report issues, and the
[Plex forums](http://forums.plexapp.com/index.php/topic/33696-proof-of-concept-plex-with-kinect/)
for feedback / discussion.

The source code of Handyplex is available on Github:
https://github.com/jopdeklein/Handyplex


CHANGES
Version 0.11 (2012/01/03)
-------------------------------------------------------------------------------
    * Added functionality to launch OSCeleton when Handyplex starts, and exit
      gracefully when Handyplex is terminated.
    * Slight tweaks for quicker repeat actions.

Version 0.1 (2011/12/01)
-------------------------------------------------------------------------------
    * Initial version


tests/gesturedetector/test_holddetector.py
"""
    Handyplex
    Copyright (C) 2011 Jop de Klein

    This file is part of Handyplex.

    Handyplex is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Handyplex is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Handyplex.  If not, see <http://www.gnu.org/licenses/>.
"""

import unittest

import sys, os.path
sys.path.append(os.path.abspath('../..'))

from handyplex.gesturedetector.holddetector import *


def do_move(gd, points):
    for p in points:
        if not gd.move(p):
            return False

    return True


class TestGestureDetector(unittest.TestCase):
    def setUp(self):
        self.callback_called = False

        def callback():
            print 'callback'
            self.callback_called = True
        self.callback = callback

    def before(self):
        gd.reset()

    def test_hold_gesture(self):
        self.gd = HoldDetector(
            gesture_callback=self.callback,
            threshold_x=50,
            threshold_y=50,
            threshold_z=50,
            min_frame_count=10,
        )

        move_ok = do_move(
            self.gd,
            [
                Point(-10, 0, 0),
                Point(-20, 20, 0),
                Point(-30, 40, 23),
                Point(30, -23, 2),
                Point(20, 03, 40),
                Point(10, 0, 20),
                Point(-30, 40, 30),
                Point(-10, 0, 20),
                Point(-30, 20, 0),
                Point(-20, 0, 40),
                Point(-30, 0, 20),
            ]
        )

        self.assertTrue(move_ok)
        self.assertTrue(self.callback_called)

    def test_hold_gesture_from_start_point(self):
        self.gd = HoldDetector(
            gesture_callback=self.callback,
            threshold_x=50,
            threshold_y=50,
            threshold_z=50,
            min_frame_count=10,
        )

        move_ok = do_move(
            self.gd,
            [
                Point(-110, 10, 0),
                Point(-120, 10, 0),
                Point(-130, 10, 0),
                Point(-140, 10, 0),
                Point(-120, 50, 40),
                Point(-90, 50, 20),
                Point(-120, 40, 30),
                Point(-110, 30, 20),
                Point(-130, 10, 0),
                Point(-120, -30, 40),
                Point(-130, -40, 20),
            ]
        )

        self.assertTrue(move_ok)
        self.assertTrue(self.callback_called)

    def test_hold_gesture_cancelled(self):
        self.gd = HoldDetector(
            gesture_callback=self.callback,
            threshold_x=50,
            threshold_y=50,
            threshold_z=50,
            min_frame_count=10,
        )

        move_ok = do_move(
            self.gd,
            [
                Point(-10, 0, 0),
                Point(-20, 20, 0),
                Point(-330, 40, 23),
                Point(30, -23, 2),
                Point(20, 03, 40),
                Point(10, 0, 20),
                Point(-30, 40, 30),
                Point(-10, 0, 20),
                Point(-30, 20, 0),
                Point(-20, 0, 40),
                Point(-30, 0, 20),
            ]
        )

        self.assertFalse(move_ok)
        self.assertFalse(self.callback_called)


if __name__ == '__main__':
    unittest.main()


tests/gesturedetector/test_swipedetector.py
"""
    Handyplex
    Copyright (C) 2011 Jop de Klein

    This file is part of Handyplex.

    Handyplex is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Handyplex is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Handyplex.  If not, see <http://www.gnu.org/licenses/>.
"""

import unittest
from OSCeleton import Point

import sys, os.path
sys.path.append(os.path.abspath('../..'))

from handyplex.gesturedetector.swipedetector import *

def do_move(gd, points):
    for p in points:
        if not gd.move(p):
            return False

    return True


class TestSwipeDetector(unittest.TestCase):

    def setUp(self):
        self.callback_called = {}

        def callback(gesture, last_point=Point(0, 0, 0)):
            self.callback_called = {
                "gesture": gesture,
                "last_point": last_point,
            }

        self.gd = SwipeDetector(
            gesture_callback=callback,
            threshold_x=50,
            threshold_y=50,
            threshold_z=50,
        )

    def before(self):
        gd.reset()

    def test_swipe_left_gesture(self):
        move_ok = do_move(
            self.gd,
            [
                Point(-10, 0, 0),
                Point(-20, 0, 0),
                Point(-40, 0, 0),
                Point(-80, 0, 0),
                Point(-120, 0, 0),
                Point(-210, 0, 0),
                Point(-220, 0, 0),
                Point(-240, 0, 0),
                Point(-280, 0, 0),
                Point(-220, 0, 0),
                Point(-320, 0, 0),
            ]
        )

        self.assertTrue(move_ok)
        self.assertEquals(self.callback_called['gesture'], 'swipe_left')
        self.assertEquals(self.callback_called['last_point'], Point(
            -220, 0, 0))

    def test_swipe_right_gesture(self):
        move_ok = do_move(
            self.gd,
            [
                Point(10, 0, 0),
                Point(20, 0, 0),
                Point(40, 0, 0),
                Point(30, 0, 0),
                Point(120, 0, 0),
                Point(210, 0, 0),
                Point(220, 0, 0),
                Point(240, 0, 0),
                Point(280, 0, 0),
                Point(220, 0, 0),
                Point(320, 0, 0),
            ]
        )

        self.assertTrue(move_ok)
        self.assertEquals(self.callback_called['gesture'], 'swipe_right')

    def test_swipe_down_gesture(self):
        move_ok = do_move(
            self.gd,
            [
                Point(0, 10, 0),
                Point(0, 20, 0),
                Point(0, 40, 0),
                Point(0, 80, 0),
                Point(0, 120, 0),
                Point(0, 210, 0),
                Point(0, 220, 0),
                Point(0, 240, 0),
                Point(0, 280, 0),
                Point(0, 220, 0),
                Point(0, 320, 0),
            ]
        )

        self.assertTrue(move_ok)
        self.assertEquals(self.callback_called['gesture'], 'swipe_down')

    def test_swipe_up_gesture(self):
        move_ok = do_move(
            self.gd,
            [
                Point(0, -10, 0),
                Point(0, -20, 0),
                Point(0, -40, 0),
                Point(0, -80, 0),
                Point(0, -120, 0),
                Point(0, -210, 0),
                Point(0, -220, 0),
                Point(0, -240, 0),
                Point(0, -280, 0),
                Point(0, -220, 0),
                Point(0, -320, 0),
            ]
        )

        self.assertTrue(move_ok)
        self.assertEquals(self.callback_called['gesture'], 'swipe_up')

    def test_swipe_pull_gesture(self):
        move_ok = do_move(
            self.gd,
            [
                Point(0, 0, 10),
                Point(0, 0, 20),
                Point(0, 0, 40),
                Point(0, 0, 80),
                Point(0, 0, 120),
                Point(0, 0, 210),
                Point(0, 0, 220),
                Point(0, 0, 240),
                Point(0, 0, 280),
                Point(0, 0, 220),
                Point(0, 0, 320),
            ]
        )

        self.assertTrue(move_ok)
        self.assertEquals(self.callback_called['gesture'], 'pull')

    def test_swipe_push_gesture(self):
        move_ok = do_move(
            self.gd,
            [
                Point(0, 0, -10),
                Point(0, 0, -20),
                Point(0, 0, -40),
                Point(0, 0, -80),
                Point(0, 0, -120),
                Point(0, 0, -210),
                Point(0, 0, -220),
                Point(0, 0, -240),
                Point(0, 0, -280),
                Point(0, 0, -220),
                Point(0, 0, -320),
            ]
        )

        self.assertTrue(move_ok)
        self.assertEquals(self.callback_called['gesture'], 'push')

    def test_fail_swipe(self):
        move_ok = do_move(
            self.gd,
            [
                Point(0, 0, 10),
                Point(0, 0, 20),
                Point(0, 20, 50),
                Point(0, 70, 90),
                Point(0, 90, 110),
                Point(0, 0, 110),
                Point(0, 0, 110),
                Point(0, 0, 110),
                Point(0, 0, 110),
                Point(0, 0, 110),
                Point(0, 0, 110),

            ]
        )

        self.assertFalse(move_ok)
        self.assertEquals(self.callback_called, {})

if __name__ == '__main__':
    unittest.main()


tests/plexcontroller/test_plexcontroller.py
"""
    Handyplex
    Copyright (C) 2011 Jop de Klein

    This file is part of Handyplex.

    Handyplex is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Handyplex is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Handyplex.  If not, see <http://www.gnu.org/licenses/>.
"""

import unittest

import sys, os.path
sys.path.append(os.path.abspath('../..'))

# FIXME: fix test

# from handyplex.plexcontroller.plexcontroller import *

# class TestPlexController(unittest.TestCase):
#     def setUp(self):
#         pass

#     def before(self):
#         pass

#     def test_initialization(self):
#         pc = PlexController(
#             'ip', 'client_name')

#         pc.detect_mode()

#         self.assertTrue(False)


# if __name__ == '__main__':
#     unittest.main()


gesturedetector/holddetector.py
"""
    Handyplex
    Copyright (C) 2011 Jop de Klein

    This file is part of Handyplex.

    Handyplex is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Handyplex is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Handyplex.  If not, see <http://www.gnu.org/licenses/>.
"""

import sys
from OSCeleton import Point

from gesturedetector import GestureDetector

class HoldDetector(GestureDetector):
    """
    Detect whether user is holding hand still on all axes for
    minimum amount of frames.
    """
    def __init__(self, gesture_callback, threshold_x=50,
            threshold_y=50, threshold_z=50, min_frame_count=200):
        self.cache = []
        self.threshold_x = threshold_x
        self.threshold_y = threshold_y
        self.threshold_z = threshold_z
        self.gesture_callback = gesture_callback
        self.min_frame_count = min_frame_count

    def _detect_gesture(self):
        if len(self.cache) == 1:
            return True

        start = self.cache[0]
        end = self.cache[-1]

        delta = end - start

        abs_delta_x = abs(delta.x)
        abs_delta_y = abs(delta.y)
        abs_delta_z = abs(delta.z)

        def _gesture_detected():
            self.gesture_callback()
            self.reset()

        if not self._is_within_cancel_threshold(abs_delta_x, abs_delta_y,
                abs_delta_z):
            print 'Hold out of bounds'
            self.reset()
            return False

        if self._get_frame_count() >= self.min_frame_count:
            _gesture_detected()

        return True

    def _is_within_cancel_threshold(self, x, y, z):
        return (x < self.threshold_x and y < self.threshold_y and
            z < self.threshold_z)



gesturedetector/swipedetector.py
"""
    Handyplex
    Copyright (C) 2011 Jop de Klein

    This file is part of Handyplex.

    Handyplex is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Handyplex is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Handyplex.  If not, see <http://www.gnu.org/licenses/>.
"""

import sys
from OSCeleton import Point

from gesturedetector import GestureDetector, GESTURES

class SwipeDetector(GestureDetector):
    """
    Detect basic swipe gestures in any direction (right/left/up/down and
    push/pull).
    """
    def _detect_gesture(self):
        start = self.cache[0]
        end = self.cache[-1]

        delta = end - start

        abs_delta_x = abs(delta.x)
        abs_delta_y = abs(delta.y)
        abs_delta_z = abs(delta.z)

        frame_count = self._get_frame_count()

        # check if we have enough frames to detect an axis
        if frame_count < SwipeDetector.min_axis_detect_count:
            return True

        # TODO introduce MovementDetector for detecting axis etc
        major_axis = self._get_major_axis(abs_delta_x, abs_delta_y, abs_delta_z)

        if not self._is_within_cancel_threshold(major_axis, abs_delta_x,
                abs_delta_y, abs_delta_z):
            self.reset()
            return False

        def _gesture_detected(detected_gesture):
            # print 'detected gesture %s' % (detected_gesture)
            self.gesture_callback(detected_gesture, end)
            self.reset()

        # check if we have enough frames to detect a gesture
        if frame_count < SwipeDetector.min_frame_count:
            return True
        # elif frame_count > SwipeDetector.max_frame_count:
        #     # print 'took too long'
        #     self.reset()
        #     return False

        if major_axis == 'x' and abs_delta_x > self.threshold_x:
            if delta.x > 0:
                _gesture_detected(GESTURES['SWIPE_RIGHT'])
            else:
                _gesture_detected(GESTURES['SWIPE_LEFT'])

        if major_axis == 'y' and abs_delta_y > self.threshold_y:
            if delta.y > 0:
                _gesture_detected(GESTURES['SWIPE_DOWN'])
            else:
                _gesture_detected(GESTURES['SWIPE_UP'])

        if major_axis == 'z' and abs_delta_z > self.threshold_z:
            if delta.z > 0:
                _gesture_detected(GESTURES['PULL'])
            else:
                _gesture_detected(GESTURES['PUSH'])

        return True


plexcontroller/plexcontroller.py
"""
    Handyplex
    Copyright (C) 2011 Jop de Klein

    This file is part of Handyplex.

    Handyplex is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Handyplex is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Handyplex.  If not, see <http://www.gnu.org/licenses/>.
"""

import sys
from gesturedetector.swipedetector import GESTURES
import httplib
import re


class PlexController:
    """
    Map gestures to commands which are sent to Plex client via HTTP API.
    Send different commands for navigation and playback mode.
    """
    xmbc_port = 3000
    plex_port = 32400

    def __init__(self, server, client):
        self.server = server
        self.client = client
        self.mode = 'nav'
        self.is_playing = False
        try:
            self._detect_mode()
        except EnvironmentError:
            print ("ERROR: Could not connect to Plex, make sure Plex "
                   "Media Server and Plex Client are running. Verify that "
                   "settings in settings.py are correct.")
            # sys.exit(1)

    def perform_gesture_action(self, gesture):
        """
        Send appropriate command for the gesture, depending on whether Plex is
        playing media or is navigating menus.
        """
        self._detect_mode()

        # TODO refactor to map from settings
        if self.mode == 'nav':
            if gesture == GESTURES['SWIPE_LEFT']:
                self._send_navigation_command('moveLeft')
            elif gesture == GESTURES['SWIPE_RIGHT']:
                self._send_navigation_command('moveRight')
            elif gesture == GESTURES['SWIPE_UP']:
                self._send_navigation_command('moveUp')
            elif gesture == GESTURES['SWIPE_DOWN']:
                self._send_navigation_command('moveDown')
            elif gesture == GESTURES['PUSH']:
                self._send_navigation_command('select')
            elif gesture == GESTURES['PULL']:
                self._send_navigation_command('back')
        elif self.mode == 'play':
            if gesture == GESTURES['SWIPE_LEFT']:
                self._send_playback_command('rewind')
            elif gesture == GESTURES['SWIPE_RIGHT']:
                self._send_playback_command('fastForward')
            elif gesture == GESTURES['SWIPE_UP']:
                self._send_application_command('sendKey?code=115')
            elif gesture == GESTURES['SWIPE_DOWN']:
                self._send_application_command('sendKey?code=116')
            elif gesture == GESTURES['PUSH']:
                if not self.is_playing:
                    self._send_playback_command('play')
                else:
                    self._send_playback_command('pause')
            elif gesture == GESTURES['PULL']:
                self._send_playback_command('stop')

    def _detect_mode(self):
        """
        Detect whether Plex is playing media, or is in navigation mode.
        """
        # TODO: find a more sustainable way of detecting mode
        path = '/xbmcCmds/xbmcHttp?command=GetCurrentlyPlaying'

        conn = httplib.HTTPConnection(self.server + ':' + str(self.xmbc_port))
        conn.request('GET', path)
        response = conn.getresponse().read()

        if response == '' or re.search(
                'Filename:\[Nothing Playing\]', response):
            self.mode = 'nav'
        else:
            self.mode = 'play'
            if re.search('PlayStatus:Playing', response):
                self.is_playing = True
            elif re.search('PlayStatus:Paused', response):
                self.is_playing = False

    def _send_navigation_command(self, command):
        self._send_command('navigation', command)

    def _send_playback_command(self, command):
        self._send_command('playback', command)

    def _send_application_command(self, command):
        self._send_command('application', command)

    def _send_command(self, controller, command):
        """
        Send command via Plex HTTP API
        """
        # TODO implement support for sending key commands directly rather than 
        # over HTTP
        path = '/system/players/' + self.client + '/'
        path += controller + '/' + command

        conn = httplib.HTTPConnection(self.server + ':' + str(self.plex_port))
        conn.request('GET', path)
        conn.getresponse()



handyplex.py
"""
    Handyplex
    Copyright (C) 2011 Jop de Klein

    This file is part of Handyplex.

    Handyplex is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Handyplex is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Handyplex.  If not, see <http://www.gnu.org/licenses/>.
"""

import sys
import settings
import signal
import os
from subprocess import Popen
from OSCeleton import OSCeleton, Skeleton, LEFT_HAND
from gesturedetector.swipedetector import SwipeDetector
from gesturedetector.movementdetector import MovementDetector
from gesturedetector.holddetector import HoldDetector
from plexcontroller.plexcontroller import PlexController


has_audio_support = False # whether wxPython is available

try:
    import wx
    has_audio_support = True
except ImportError:
    pass

osceleton = None # Handle to osceleton process

last_message = ''
def print_once(msg):
    """
    Simple helper function for printing a message only once, handy for use in
    the main_loop.
    """
    global last_message
    if last_message != msg:
        last_message = msg
        print msg


class Handyplex:
    """
    Retrieve Kinect info via OSCeleton, detecting user's hands and
    subsequently recognize gestures and send commands to Plex.
    """
    def __init__(self, plex_server, plex_client):
        """
        Establish OSCeleton connection and initialize PlexController and
        Gesture Detectors.
        """
        print "Initializing..."
        self.server = OSCeleton(7110)
        self.frame_count = 0
        """List of users / hands, currently only supporting one concurrently"""
        self.users = {}

        # FIXME: check IP/Client
        self.pc = PlexController(plex_server, plex_client)

        # Set up GestureDetector instances. Since we currently support one
        # hand / user at a time we can initialize all at once.
        gt = settings.THRESHOLDS['GESTURE']
        self.sd = SwipeDetector(self._gesture_detected,
            threshold_x=gt['x'],
            threshold_y=gt['y'],
            threshold_z=gt['z'],
            threshold_cancel_factor=gt['cancel_factor'])

        mt = settings.THRESHOLDS['MOVEMENT']
        self.md = MovementDetector(self._move_detected,
            threshold_x=mt['x'],
            threshold_y=mt['y'],
            threshold_z=mt['z'],
            threshold_cancel_factor=mt['cancel_factor'])

        ht = settings.THRESHOLDS['HOLD']
        self.hd = HoldDetector(self._hold_detected, 
            threshold_x=ht['x'],
            threshold_y=ht['y'],
            threshold_z=ht['z'],
            min_frame_count=12)

        self.repeat_frame_count = 0
        self.repeat_count = 0

        self._set_state('wait_for_hold')

        print 'Done initializing'
        while True:
            self._main_loop()

    def _main_loop(self):
        """
        Poll the OSCeleton server for frames. Detects new/lost user and relays
        Kinect hand data to appropriate GestureDetector based on current
        application state.
        """
        is_ok = self.server.run()
        if not is_ok:
            self._lost_user()

        if self.server.frames > self.frame_count:

            for user in self.server.get_new_skeletons():
                if user.id not in self.users:
                    self._new_user(user)

                l_hand = user[LEFT_HAND]
                if l_hand:
                    if self.state == 'wait_for_hold':
                        print_once('waiting for hold...')
                        self.hd.move(l_hand)
                    elif self.state == 'wait_for_gesture':
                        print_once('detecting gesture...')
                        move_ok = self.sd.move(l_hand)
                        if not move_ok:
                            self._set_state('wait_for_hold')
                    elif self.state == 'wait_for_repeat':
                        print_once('waiting for repeat...')
                        move_ok = self.md.move(l_hand)

                        if move_ok == False:
                            print_once('repeat cancelled')
                            self.repeat_frame_count = 0
                            self._set_state('wait_for_hold')

            self.frame_count += 1
            return True

    def _gesture_detected(self, gesture, last_point):
        """
        Sends gesture command to PlexController and prepares for repeat
        gesture if gesture was performed on x or y axis.
        """
        print 'gesture detected: %s' % (gesture)
        self.pc.perform_gesture_action(gesture)
        self.sd.reset()

        # currently only needs repeat on x and y axis
        direction = None
        if gesture == 'swipe_left':
            direction = 'left'
        elif gesture == 'swipe_right':
            direction = 'right'
        elif gesture == 'swipe_up':
            direction = 'up'
        elif gesture == 'swipe_down':
            direction = 'down'

        if direction:
            self._set_state('wait_for_repeat', direction)
            self.last_gesture = gesture
        else:
            self._set_state('wait_for_hold')

    def _hold_detected(self):
        """Hold detected, transition to wait_for_gesture state"""
        print 'hold detected'
        self._set_state('wait_for_gesture')

    def _move_detected(self, delta):
        """
        Potential repeat movement detected, repeat when repeat conditions
        apply.
        """
        self.repeat_frame_count = self.repeat_frame_count + 1
        # TODO: make min_repeat_count configurable, or introduce
        # RepeatDetector
        if self.repeat_frame_count > 10:
            self._perform_repeat(delta)

    def _perform_repeat(self, delta):
        """
        Repeatedly perform previous action, in intervals based on distance of
        previous move.
        """
        self.repeat_count = self.repeat_count + 1
        delta = abs(delta)

        # TODO use some math :)
        if delta < 0.02:
            repeat_interval = 50
        if delta < 0.05:
            repeat_interval = 40
        elif delta < 0.1:
            repeat_interval = 30
        elif delta < 0.15:
            repeat_interval = 20
        elif delta < 0.20:
            repeat_interval = 10
        elif delta < 0.40:
            repeat_interval = 3
        else:
            repeat_interval = 1

        if self.repeat_count > repeat_interval:
            self.pc.perform_gesture_action(self.last_gesture)
            self.repeat_count = 0

    def _new_user(self, user):
        print 'new user detected: %s' % (user.id)
        self.users[user.id] = Skeleton(user.id)
        self.sd.reset()

        self._play_sound(settings.SOUNDS['NEW_USER'])

    def _lost_user(self):
        print 'lost user'
        self.users = {} # FIXME now supporting only one user/hand at a time
        self._play_sound(settings.SOUNDS['LOST_USER'])

    def _set_state(self, state, direction=None):
        """
        Initialize appropriate GestureDetectors based on state.
        Currently supports the following states:
            wait_for_hold
            wait_for_gesture
            wait_for_repeat
        """
        self.state = state
        if state == 'wait_for_hold':
            self.md.reset()
            self.hd.reset()
        elif state == 'wait_for_gesture':
            self.sd.reset()
        elif state == 'wait_for_repeat':
            self.md.reset(direction)

    def _play_sound(self, sound_path):
        if has_audio_support:
            sound = wx.Sound(sound_path)
            sound.Play(wx.SOUND_SYNC)


def cleanup():
    """Graceful exit if application quits"""
    global osceleton
    print 'Exiting, thanks for using Handyplex!'
    osceleton.terminate()
    sys.exit(0)

def signal_handler(signal, frame):
    cleanup()

signal.signal(signal.SIGINT, signal_handler)

def except_hook(type, value, tb):
    cleanup()

sys.excepthook = except_hook

def main(argv):
    global osceleton
    if has_audio_support:
        wx.PySimpleApp()

    # Call OSCeleton
    osceleton = Popen([os.path.join(settings.OSCELETON_PATH, 'osceleton'),
        '-n', '-f'])

    p = Handyplex(settings.PLEX_SERVER_IP, settings.PLEX_CLIENT_NAME)

if __name__ == "__main__":
    main(sys.argv)
